File: apps/proxy/internal/auth/cache.go
package auth

import (
	"context"
	"encoding/json"
	"time"
)

func validationKey(keyHash string) string { return validationPrefix + ":" + keyHash }

func lastUsedKey(apiKeyID string) string { return lastUsedPrefix + ":" + apiKeyID }

func disabledModelsKey(userID string) string { return disabledModelsPrefix + ":" + userID }

func analyticsVersionKey(userID string) string { return analyticsVersionPrefix + ":" + userID }

func analyticsVersionBumpKey(userID string) string { return analyticsVersionBumpPref + ":" + userID }

func (s *Service) getCachedAPIKeyValidation(ctx context.Context, keyHash string) (cacheValue, bool) {
	raw, err := s.redis.Get(ctx, validationKey(keyHash)).Result()
	if err != nil || raw == "" {
		return cacheValue{}, false
	}
	var value cacheValue
	if err := json.Unmarshal([]byte(raw), &value); err != nil {
		return cacheValue{}, false
	}
	return value, true
}

func (s *Service) setCachedAPIKeyValidation(ctx context.Context, keyHash string, value cacheValue, ttl time.Duration) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return s.redis.Set(ctx, validationKey(keyHash), data, ttl).Err()
}

func (s *Service) InvalidateAPIKeyValidation(ctx context.Context, keyHash, apiKeyID string) error {
	keys := []string{validationKey(keyHash)}
	if apiKeyID != "" {
		keys = append(keys, lastUsedKey(apiKeyID))
	}
	return s.redis.Del(ctx, keys...).Err()
}

func (s *Service) touchAPIKeyLastUsed(ctx context.Context, apiKeyID string) {
	if apiKeyID == "" {
		return
	}
	updated, err := s.redis.SetNX(ctx, lastUsedKey(apiKeyID), "1", lastUsedTTL).Result()
	if err == nil && !updated {
		return
	}
	_, _ = s.db.NewUpdate().TableExpr("proxy_api_key").Set("\"lastUsedAt\" = NOW()").Where("id = ?", apiKeyID).Exec(ctx)
}

func (s *Service) getCachedDisabledModels(ctx context.Context, userID string) ([]string, bool) {
	raw, err := s.redis.Get(ctx, disabledModelsKey(userID)).Result()
	if err != nil || raw == "" {
		return nil, false
	}
	var value disabledModelsCacheValue
	if err := json.Unmarshal([]byte(raw), &value); err != nil {
		return nil, false
	}
	return s.normalizeModelList(value.Models), true
}

func (s *Service) setCachedDisabledModels(ctx context.Context, userID string, modelList []string) error {
	data, err := json.Marshal(disabledModelsCacheValue{Models: s.normalizeModelList(modelList)})
	if err != nil {
		return err
	}
	return s.redis.Set(ctx, disabledModelsKey(userID), data, disabledModelsTTL).Err()
}

func (s *Service) BumpAnalyticsCacheVersionThrottled(ctx context.Context, userID string) {
	if userID == "" {
		return
	}
	updated, err := s.redis.SetNX(ctx, analyticsVersionBumpKey(userID), "1", analyticsBumpTTL).Result()
	if err != nil || !updated {
		return
	}
	version, err := s.redis.Incr(ctx, analyticsVersionKey(userID)).Result()
	if err == nil && version == 1 {
		_ = s.redis.Expire(ctx, analyticsVersionKey(userID), analyticsVersionTTL).Err()
	}
}

File: apps/proxy/src/auth/cache.ts
import type Redis from "ioredis";
import type { CacheValue, DisabledModelsCacheValue, RateLimitRule } from "./types.js";

export const validationPrefix = "opendum:api-key:validation";
export const lastUsedPrefix = "opendum:api-key:last-used";
export const disabledModelsPrefix = "opendum:user:disabled-models";

export const validTTL = 45 * 1000;
export const invalidTTL = 10 * 1000;
export const lastUsedTTL = 60 * 1000;
export const disabledModelsTTL = 60 * 1000;
export const analyticsVersionPrefix = "opendum:analytics:v1:version";
export const analyticsVersionBumpPrefix = "opendum:analytics:v1:version-bump";
export const analyticsVersionTTL = 30 * 24 * 3600 * 1000;
export const analyticsBumpTTL = 15 * 1000;

export function validationKey(keyHash: string): string {
  return `${validationPrefix}:${keyHash}`;
}

export function lastUsedKey(apiKeyId: string): string {
  return `${lastUsedPrefix}:${apiKeyId}`;
}

export function disabledModelsKey(userId: string): string {
  return `${disabledModelsPrefix}:${userId}`;
}

export function analyticsVersionKey(userId: string): string {
  return `${analyticsVersionPrefix}:${userId}`;
}

export function analyticsVersionBumpKey(userId: string): string {
  return `${analyticsVersionBumpPrefix}:${userId}`;
}

export async function getCachedAPIKeyValidation(redis: Redis | null, keyHash: string): Promise<CacheValue | null> {
  if (!redis) return null;
  try {
    const raw = await redis.get(validationKey(keyHash));
    if (!raw) return null;
    return JSON.parse(raw) as CacheValue;
  } catch {
    return null;
  }
}

export async function setCachedAPIKeyValidation(redis: Redis | null, keyHash: string, value: CacheValue, ttlMs: number): Promise<void> {
  if (!redis) return;
  try {
    await redis.set(validationKey(keyHash), JSON.stringify(value), "PX", ttlMs);
  } catch {
    // ignore
  }
}

export async function invalidateAPIKeyValidation(redis: Redis | null, keyHash: string, apiKeyId: string): Promise<void> {
  if (!redis) return;
  const keys = [validationKey(keyHash)];
  if (apiKeyId !== "") keys.push(lastUsedKey(apiKeyId));
  try {
    await redis.del(...keys);
  } catch {
    // ignore
  }
}

export async function getCachedDisabledModels(redis: Redis | null, userId: string): Promise<string[] | null> {
  if (!redis) return null;
  try {
    const raw = await redis.get(disabledModelsKey(userId));
    if (!raw) return null;
    const value = JSON.parse(raw) as DisabledModelsCacheValue;
    return value.models ?? null;
  } catch {
    return null;
  }
}

export async function setCachedDisabledModels(redis: Redis | null, userId: string, modelList: string[]): Promise<void> {
  if (!redis) return;
  try {
    await redis.set(disabledModelsKey(userId), JSON.stringify({ models: modelList } satisfies DisabledModelsCacheValue), "PX", disabledModelsTTL);
  } catch {
    // ignore
  }
}

export async function bumpAnalyticsCacheVersionThrottled(redis: Redis | null, userId: string): Promise<void> {
  if (!redis || userId === "") return;
  try {
    const bumpKey = analyticsVersionBumpKey(userId);
    const updated = await redis.set(bumpKey, "1", "PX", analyticsBumpTTL, "NX");
    if (updated !== "OK") return;
    const version = await redis.incr(analyticsVersionKey(userId));
    if (version === 1) {
      await redis.expire(analyticsVersionKey(userId), analyticsVersionTTL / 1000);
    }
  } catch {
    // ignore
  }
}

export function serializeRateLimitRules(rules: RateLimitRule[]): RateLimitRule[] {
  return rules;
}
